### ISSUE TITLE
Azure File Copy stopped working

### ISSUE THREAD

**Issue by ericlemes** [microsoft/azure-pipelines-tasks#13774]
## Required Information

Entering this information will route you directly to the right team and expedite traction.

**Question, Bug, or Feature?**  
*Type*: Bug

**Enter Task Name**: AzureFileCopyV2

## Environment
- Server - Azure Pipelines or TFS on-premises?
    
    - If using Azure Pipelines, provide the account name, team project name, build definition name/build number: 
Account name: squk-central
Team: Code Metrics
Build definition: Release definition 2
Build id: release id 195

https://squk-central.visualstudio.com/Code%20Metrics/_releaseProgress?_a=release-environment-logs&releaseId=195&environmentId=363

- Agent - Hosted or Private: Private.

    - If using private agent, provide the OS of the machine running the agent and the agent version: Windows 10, 2.174.1

## Issue Description

I have a production release pipeline on Azure DevOps with AzureFileCopyV2. It stopped working without apparent reason.

I followed a lot of posts regarding permissions to service principal, and it seems that the content is being uploaded to the storage account but by whatever reason (the log isn't clear), it is unable to download on the target machine.

### Task logs

2020-10-26T15:07:52.1182836Z ##[section]Starting: Copy Files to Azure VM
2020-10-26T15:07:52.1821762Z ==============================================================================
2020-10-26T15:07:52.1822447Z Task         : Azure file copy
2020-10-26T15:07:52.1823085Z Description  : Copy files to Azure Blob Storage or virtual machines
2020-10-26T15:07:52.1823692Z Version      : 2.175.3
2020-10-26T15:07:52.1824213Z Author       : Microsoft Corporation
2020-10-26T15:07:52.1824964Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/deploy/azure-file-copy
2020-10-26T15:07:52.1825745Z ==============================================================================
2020-10-26T15:07:56.6443324Z Added TLS 1.2 in session.
2020-10-26T15:07:57.6998914Z ##[command]Import-Module -Name C:\Program Files\WindowsPowerShell\Modules\AzureRM\6.13.1\AzureRM.psd1 -Global
2020-10-26T15:08:29.0523994Z ##[command]Clear-AzureRmContext -Scope Process
2020-10-26T15:08:29.0525091Z ##[command]Disable-AzureRmContextAutosave -ErrorAction Stop
2020-10-26T15:08:31.3756886Z WARNING: AzureRM.Profile already loaded. Az and AzureRM modules cannot be imported in the same session or used in the 
2020-10-26T15:08:31.3758369Z same script or runbook. If you are running PowerShell in an environment you control you can use the 'Uninstall-AzureRm'
2020-10-26T15:08:31.3759551Z  cmdlet to remove all AzureRm modules from your machine. If you are running in Azure Automation, take care that none of
2020-10-26T15:08:31.3760555Z  your runbooks import both Az and AzureRM modules. More information can be found here: 
2020-10-26T15:08:31.3761358Z https://aka.ms/azps-migration-guide.
2020-10-26T15:08:32.1313383Z WARNING: AzureRM.Profile already loaded. Az and AzureRM modules cannot be imported in the same session or used in the 
2020-10-26T15:08:32.1315177Z same script or runbook. If you are running PowerShell in an environment you control you can use the 'Uninstall-AzureRm'
2020-10-26T15:08:32.1316861Z  cmdlet to remove all AzureRm modules from your machine. If you are running in Azure Automation, take care that none of
2020-10-26T15:08:32.1318449Z  your runbooks import both Az and AzureRM modules. More information can be found here: 
2020-10-26T15:08:32.2241608Z https://aka.ms/azps-migration-guide.
2020-10-26T15:08:32.6757734Z ##[command]Add-AzureRMAccount -ServicePrincipal -Tenant *** -Credential System.Management.Automation.PSCredential -Environment AzureCloud @processScope
2020-10-26T15:08:34.4633289Z ##[command] Select-AzureRMSubscription -SubscriptionId c25996f4-001e-4393-9d9b-705d6baef5a7 -TenantId ***
2020-10-26T15:08:43.2132841Z Uploading files from source path: 'C:\agent\_work\r1\a\Code Metrics Latest' to storage account: 'sonarqubediag174' in container: '6fc05552-86b8-42f1-ad73-fdba84a22cc2' with blob prefix: ''
2020-10-26T15:08:43.2381933Z ##[command] & "AzCopy\AzCopy.exe" /Source:"C:\agent\_work\r1\a\Code Metrics Latest" /Dest:"https://sonarqubediag174.blob.core.windows.net/6fc05552-86b8-42f1-ad73-fdba84a22cc2" /@:"C:\agent\_work\_temp\d5ndlb21nwprdja5fpkh3vunmi" /XO /Y /SetContentType /Z:"AzCopy" /V:"AzCopy\AzCopyVerbose_b55edc96-a4b0-4ec9-b989-dfe89c67dbfb.log" /S
2020-10-26T15:09:26.2329960Z [2020/10/26 08:09:26] Transfer summary:
2020-10-26T15:09:26.2330737Z -----------------
2020-10-26T15:09:26.2331218Z Total files transferred: 1571
2020-10-26T15:09:26.2331719Z Transfer successfully:   1571
2020-10-26T15:09:26.2332187Z Transfer skipped:        0
2020-10-26T15:09:26.2332611Z Transfer failed:         0
2020-10-26T15:09:26.2333141Z Elapsed time:            00.00:00:42
2020-10-26T15:09:26.7168542Z Uploaded files successfully from source path: 'C:\agent\_work\r1\a\Code Metrics Latest' to storage account: 'sonarqubediag174' in container: '6fc05552-86b8-42f1-ad73-fdba84a22cc2' with blob prefix: ''
2020-10-26T15:09:36.4100595Z [Azure Call]Getting the custom script extension 'WinRMCustomScriptExtension' for vm 'SonarQube-App'
2020-10-26T15:09:36.8142781Z [Azure Call]Got the custom script extension 'WinRMCustomScriptExtension' for vm 'SonarQube-App'
2020-10-26T15:09:36.8605371Z [Azure Call]Getting the status for vm 'SonarQube-App'
2020-10-26T15:09:37.7224961Z [Azure Call]Got the status for vm 'SonarQube-App'
2020-10-26T15:09:42.6212623Z PSSession created for Machines:'codemetrics.westeurope.cloudapp.azure.com'
2020-10-26T15:09:44.7497727Z ================================================ codemetrics.westeurope.cloudapp.azure.com ================================================
2020-10-26T15:09:49.8348547Z ================================================ codemetrics.westeurope.cloudapp.azure.com ================================================
2020-10-26T15:09:49.9243233Z ##[command]& 'C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe' -NoLogo -NoProfile -NonInteractive 
2020-10-26T15:09:49.9244190Z -ExecutionPolicy Unrestricted -Command ". 
2020-10-26T15:09:49.9244941Z 'C:\Users\SonarQube\AppData\Local\Temp\2f0050b2-7ff7-4484-b0b3-90aa10348e60.ps1'"
2020-10-26T15:09:49.9245737Z 
2020-10-26T15:09:49.9290741Z 
2020-10-26T15:09:49.9292790Z 
2020-10-26T15:09:49.9325384Z 
2020-10-26T15:09:49.9327504Z 
2020-10-26T15:09:49.9357930Z     Directory: C:\Windows\DtlDownloads\AFC_48c7c1a7-6989-4dbb-bac0-0dfe2f3d5ea3
2020-10-26T15:09:49.9358804Z 
2020-10-26T15:09:49.9389152Z 
2020-10-26T15:09:49.9391268Z 
2020-10-26T15:09:49.9418876Z 
2020-10-26T15:09:49.9420919Z 
2020-10-26T15:09:49.9447853Z Mode                LastWriteTime         Length Name                                             
2020-10-26T15:09:49.9448767Z 
2020-10-26T15:09:49.9480041Z ----                -------------         ------ ----                                             
2020-10-26T15:09:49.9481056Z 
2020-10-26T15:09:49.9511908Z d-----       10/26/2020   3:09 PM                AzCopy                                           
2020-10-26T15:09:49.9512664Z 
2020-10-26T15:09:54.9588372Z ================================================ codemetrics.westeurope.cloudapp.azure.com ================================================
2020-10-26T15:09:59.9731806Z ================================================ codemetrics.westeurope.cloudapp.azure.com ================================================
2020-10-26T15:10:04.9734768Z ================================================ codemetrics.westeurope.cloudapp.azure.com ================================================
2020-10-26T15:10:09.9813188Z ================================================ codemetrics.westeurope.cloudapp.azure.com ================================================
2020-10-26T15:10:14.9863790Z ================================================ codemetrics.westeurope.cloudapp.azure.com ================================================
2020-10-26T15:10:20.1786887Z ================================================ codemetrics.westeurope.cloudapp.azure.com ================================================
2020-10-26T15:10:25.0214106Z ================================================ codemetrics.westeurope.cloudapp.azure.com ================================================
2020-10-26T15:10:30.0397551Z ================================================ codemetrics.westeurope.cloudapp.azure.com ================================================
2020-10-26T15:10:30.2549046Z [2020/10/26 15:10:27][ERROR] Error parsing source location 
2020-10-26T15:10:30.2551276Z "https://sonarqubediag174.blob.core.windows.net/6fc05552-86b8-42f1-ad73-fdba84a22cc2": Failed to enumerate directory 
2020-10-26T15:10:30.2553000Z https://sonarqubediag174.blob.core.windows.net/6fc05552-86b8-42f1-ad73-fdba84a22cc2 with file pattern . The underlying 
2020-10-26T15:10:30.2554457Z connection was closed: An unexpected error occurred on a receive.
2020-10-26T15:10:30.2556935Z At line:1 char:1
2020-10-26T15:10:30.2558109Z + & 'C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe' -NoLogo ...
2020-10-26T15:10:30.2559244Z + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2020-10-26T15:10:30.2560663Z     + CategoryInfo          : NotSpecified: ([2020/10/26 15:...d on a receive.:String) [], RemoteException
2020-10-26T15:10:30.2563500Z     + FullyQualifiedErrorId : NativeCommandError
2020-10-26T15:10:30.2564838Z  
2020-10-26T15:10:30.2565627Z 
2020-10-26T15:10:30.2728234Z 
2020-10-26T15:10:30.2754469Z HttpStatusMessage:
2020-10-26T15:10:30.2757686Z RequestId:
2020-10-26T15:10:30.2761016Z Time: For more details, please type "AzCopy /?:Source" or use verbose option /V.
2020-10-26T15:10:30.2763118Z 
2020-10-26T15:10:30.2766139Z 
2020-10-26T15:10:30.2792897Z 
2020-10-26T15:10:30.2824184Z 
2020-10-26T15:10:30.2854815Z 
2020-10-26T15:10:30.2856458Z 
2020-10-26T15:10:30.2885083Z 
2020-10-26T15:10:30.2886739Z 
2020-10-26T15:10:32.2347204Z WARNING: AzureRM.Profile already loaded. Az and AzureRM modules cannot be imported in the same session or used in the 
2020-10-26T15:10:32.2358061Z same script or runbook. If you are running PowerShell in an environment you control you can use the 'Uninstall-AzureRm'
2020-10-26T15:10:32.2365374Z  cmdlet to remove all AzureRm modules from your machine. If you are running in Azure Automation, take care that none of
2020-10-26T15:10:32.2366840Z  your runbooks import both Az and AzureRM modules. More information can be found here: 
2020-10-26T15:10:32.2368105Z https://aka.ms/azps-migration-guide.
2020-10-26T15:10:32.8760598Z WARNING: AzureRM.Profile already loaded. Az and AzureRM modules cannot be imported in the same session or used in the 
2020-10-26T15:10:32.8762638Z same script or runbook. If you are running PowerShell in an environment you control you can use the 'Uninstall-AzureRm'
2020-10-26T15:10:32.8764473Z  cmdlet to remove all AzureRm modules from your machine. If you are running in Azure Automation, take care that none of
2020-10-26T15:10:32.8765964Z  your runbooks import both Az and AzureRM modules. More information can be found here: 
2020-10-26T15:10:32.8773994Z https://aka.ms/azps-migration-guide.
2020-10-26T15:10:33.1581590Z ##[command]Logout-AzAccount -Scope Process -ErrorAction Stop
2020-10-26T15:10:34.0170736Z WARNING: AzureRM.Profile already loaded. Az and AzureRM modules cannot be imported in the same session or used in the 
2020-10-26T15:10:34.0172725Z same script or runbook. If you are running PowerShell in an environment you control you can use the 'Uninstall-AzureRm'
2020-10-26T15:10:34.0174382Z  cmdlet to remove all AzureRm modules from your machine. If you are running in Azure Automation, take care that none of
2020-10-26T15:10:34.0175867Z  your runbooks import both Az and AzureRM modules. More information can be found here: 
2020-10-26T15:10:34.0245685Z https://aka.ms/azps-migration-guide.
2020-10-26T15:10:34.5299516Z ##[error]Non-Zero exit code: '1' for ComputerName: 'SonarQube-App'
2020-10-26T15:10:34.9968307Z ##[section]Finishing: Copy Files to Azure VM

### Error logs

[ERROR] Error parsing source location "https://sonarqubediag174.blob.core.windows.net/6fc05552-86b8-42f1-ad73-fdba84a22cc2": Failed to enumerate directory https://sonarqubediag174.blob.core.windows.net/6fc05552-86b8-42f1-ad73-fdba84a22cc2 with file pattern . The underlying connection was closed: An unexpected error occurred on a receive.

**Comment by AmrutaKawade** [comment]
@ericlemes Can you try Azure File Copy Version 3 or 4? 4 requires some extra permissions as mentioned here https://brettmckenzie.net/2020/03/23/azure-pipelines-copy-files-task-authentication-failed/

**Comment by ericlemes** [comment]
Hi @AmrutaKawade I tried v4 and got a similar error.

https://squk-central.visualstudio.com/Code%20Metrics/_releaseProgress?_a=release-environment-logs&releaseId=194&environmentId=361

2020-10-26T14:43:04.6723176Z Failed while downloading azcopy.exe from the URL with exception Exception calling "DownloadFile"

**Comment by ericlemes** [comment]
You can ping me on Teams if you'd like to discuss in details (Alias ercintra)

**Comment by 20shivangi** [comment]
@ericlemes Can you send a snapshot of your pipeline once or provide permission to view the pipeline.. I am unable to open the link that you have provided.
While copying to VM, try to enable the options as shown in snapshot 

![image](https://user-images.githubusercontent.com/45754171/97965009-afd7cc80-1ddf-11eb-93eb-533c90c96416.png)

**Comment by ericlemes** [comment]
Hi @20shivangi,

Unfortunately I can't give you rights for this repo which is private. I've been debugging this issue with @AmrutaKawade and we found out that the issue is because this machine has TLS versions older than 1.2 disabled and the internal way of downloading the azure copy to the remote machine that will then download the files uses System.Net.WebClient which by default tries to use older TLS versions.

Hence the reason why in the log we can see:
2020-10-26T15:10:30.2554457Z connection was closed: An unexpected error occurred on a receive.

Not sure about the solution though.
